netebpfext/user/nmr_impl.cpp

```diff
@@ -0,0 +1,303 @@
+    if (NT_SUCCESS(status)) {
+        *provider_binding_context = binding.provider_binding_context;
+        *provider_dispatch = binding.provider_dispatch;
+    }
+    return status;
+}
+
+std::optional<_nmr::pending_action_t>
+_nmr::bind(client_registration& client, provider_registration& provider)
+{
+    // Match on NPI ID
+    if (*client.characteristics.ClientRegistrationInstance.NpiId !=
+        *provider.characteristics.ProviderRegistrationInstance.NpiId) {
+        return std::nullopt;
+    }
+
+    // Skip if client or provider are deregestering.
+    if (client.deregistering || provider.deregistering) {
+        return std::nullopt;
+    }
+
+    // Acquire references on both client and provider to prevent them from unloading.
+    client.bindings++;
+    provider.bindings++;
+
+    nmr_binding_handle h = reinterpret_cast<nmr_binding_handle>(next_handle++);
+    bindings.insert({h, {provider, client}});
+
+    return {[&client, &provider, h, this]() {
+        NTSTATUS status = client.characteristics.ClientAttachProvider(
+            h, client.context, &provider.characteristics.ProviderRegistrationInstance);
+        if (!NT_SUCCESS(status)) {
+            unbind_complete(h);
+        }
+    }};
+}
+
+void
+_nmr::unbind_complete(nmr_binding_handle binding_handle)
+{
+    std::unique_lock l(lock);
+    auto it = bindings.find(binding_handle);
+    if (it == bindings.end()) {
+        throw std::runtime_error("invalid handle");
+    }
+    auto& binding = it->second;
+
+    // Notify the client that that the binding context can be freed.
+    binding.client.characteristics.ClientCleanupBindingContext(binding.client_binding_context);
+
+    // Notify the provider that that the binding context can be freed.
+    binding.provider.characteristics.ProviderCleanupBindingContext(binding.provider_binding_context);
+
+    binding.provider.bindings--;
+    binding.client.bindings--;
+    bindings.erase(it);
+
+    // Notify the client or provider to check if they have any pending bindings.
+    bindings_changed.notify_all();
+}
+
+bool
+_nmr::unbind(nmr_binding_handle binding_handle)
+{
+    std::unique_lock l(lock);
+    auto it = bindings.find(binding_handle);
+    if (it == bindings.end()) {
+        throw std::runtime_error("invalid handle");
+    }
+
+    auto& binding = it->second;
+    l.unlock();
+    NTSTATUS client_detach_status = binding.client.characteristics.ClientDetachProvider(binding.client_binding_context);
+    NTSTATUS provider_detach_status =
+        binding.provider.characteristics.ProviderDetachClient(binding.provider_binding_context);
+    if (NT_SUCCESS(client_detach_status) && NT_SUCCESS(provider_detach_status)) {
+        unbind_complete(binding_handle);
+        return false;
+    }
+    return true;
+}
+
+template <typename collection_t, typename characteristics_t>
+collection_t::value_type::first_type
+_nmr::add(collection_t& collection, characteristics_t& characteristics, void* context)
```

Comment:
```suggestion
_nmr::add(_Inout_ collection_t& collection, _In_ characteristics_t& characteristics, _In_ void* context)
```
Also, should characteristics be const?